lntest/itest/network.go

```diff
@@ -0,0 +1,66 @@
+// +build rpctest
+
+package itest
+
+import (
+	"context"
+	"time"
+
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/lntest"
+	"github.com/lightningnetwork/lnd/lntest/wait"
+)
+
+// testNetworkConnectionTimeout checks that the connectiontimeout is taking
+// effect. It creates a node with a small connection timeout value, and connects
+// it to a non-routable IP address.
+func testNetworkConnectionTimeout(net *lntest.NetworkHarness, t *harnessTest) {
+	var (
+		ctxt, _ = context.WithTimeout(context.Background(), defaultTimeout)
+		// testPub is a random public key for testing only.
+		testPub = "0332bda7da70fefe4b6ab92f53b3c4f4ee7999" +
+			"f312284a8e89c8670bb3f67dbee2"
+		// testHost is a non-routable IP address. It's used to cause a
+		// connection timeout.
+		testHost = "10.255.255.1"
+	)
+
+	// First, test the global timeout settings.
+	// Create Carol with a connection timeout of 1 millisecond.
+	carol, err := net.NewNode("Carol", []string{"--connectiontimeout=1ms"})
+	if err != nil {
+		t.Fatalf("unable to create new node carol: %v", err)
+	}
+	defer shutdownAndAssert(net, t, carol)
+
+	// Try to connect Carol to a non-routable IP address, which should give us
+	// a timeout error.
+	req := &lnrpc.ConnectPeerRequest{
+		Addr: &lnrpc.LightningAddress{
+			Pubkey: testPub,
+			Host:   testHost,
+		},
+	}
+	assertTimeoutError(ctxt, t, carol, req)
+}
+
+// assertTimeoutError asserts a connection timeout error is raised. With the
+// waiting time of 15 seconds, we will not see a timeout error if the default
+// timeout value of 120 seconds is used. A custom timeout value must be set in
+// order to raise the error. 1ms is good as we can see the test result quickly,
+// while any value below 15 seconds would be working.
+func assertTimeoutError(ctxt context.Context, t *harnessTest,
+	node *lntest.HarnessNode, req *lnrpc.ConnectPeerRequest) {
+
+	err := wait.Predicate(func() bool {
+		_, err := node.ConnectPeer(ctxt, req)
```

Comment:
If connect peer blocks until we get a connection (or fail), we don't need the wait predicate here? This is more used for cases where we need multiple calls to wait until come condition is met (eg, call list channels until our channel shows up, or fail).

Reply:
Originally I was trying the test the customized `timeout` was taking effect. Because the connection will fail because of timeout no matter what, I want to make sure it's timing out using the specified value, not the default one. Now the default one uses 120 seconds, which is much smaller than 15 seconds using in the `Predicate`, and the customized value is 1ms. So if the connect peer blocks or fail, it means the customized timeout value is not used, and that's what it's testing.

Reply:
Still not entirely sure of the use of wait predicate here, since we're using it to once off check a function rather than repeatedly query a condition. But I see what you're going for, and it's probably not worth rolling our own for this specific case.

Reply:
Hadn't seen this comment before, but using a context timeout on `ConnectPeer` seems like a better approach instead of `wait.Predicate`? If we see a context deadline error, then we know the custom timeout isn't being used.